R/renv_helpers.R

```diff
@@ -0,0 +1,150 @@
+#' read_renv_lock
+#'
+#' Reads renv.lock if it exists and can be parsed as json.
+#'
+#' @param renv_lock_path location of the renv.lock file, defaults to "renv.lock"
+#'
+#' @return the result of reading renv.lock with [jsonlite::read_json]
+#'
+#' @importFrom jsonlite read_json
+read_renv_lock <- function(renv_lock_path = "renv.lock") {
+  if (!file.exists(renv_lock_path)) {
+    stop(renv_lock_path, " does not exist!")
+  }
+  tryCatch(
+    renv_lock <- jsonlite::read_json(renv_lock_path),
+    error = function(e) {
+      stop("Error reading renv.lock file\n", e)
+    }
+  )
+  renv_lock
+}
+
+#' renv_remote_pkg
+#'
+#' Construct a list to be passed in a list to the git_pkgs argument of [rix]
+#' The list returned contains the information necessary to have nix attempt to
+#' build the package from an external repository.
+#'
+#' @param renv_lock_pkg_info a the list representation of a single package
+#' entry from an renv.lock file.
+#' @param type the type of remote package, defaults to the RemoteType of the
+#' renv entry.
+#' currently supported types: 'github' 'gitlab'
+#' see [remotes](https://remotes.r-lib.org/) for more.
+#'
+#' @return a list with three elements named:
+#'  "package_name", "repo_url", "commit"
+#'
+#' @examples
+#' \dontrun{
+#' renv_remote_pkgs(read_renv_lock()$Packages$renv)
+#' }
+renv_remote_pkg <- function(
+    renv_lock_pkg_info,
+    type = renv_lock_pkg_info$RemoteType) {
+  type <- match.arg(type, c(
+    "github", "gitlab"
+    # , "bitbucket", "git", "local", "svn", "url", "version", "cran", "bioc"
+  ))
+  pkg_info <- vector(mode = "list", length = 3)
+  names(pkg_info) <- c("package_name", "repo_url", "commit")
+  switch(type,
+    "github" = {
+      pkg_info[[1]] <- renv_lock_pkg_info$Package
```

Comment:
I'm not sure how this would help reduce the repetition. The two blocks differ as github does not use `RemoteHost` and I anticipate that the different remote types may similarly need slightly different handling which is why I repeated the `Package` and `RemoteSha` assignments rather than placing them outside the switch in case one other other remotes need some different processing to get at these values as github does with `RemoteHost`.

Reply:
True, but I would prefer this logic to be handled in one function, and there would still be a lot of boilerplate that could go into that function as well.

Reply:
Do you mean a function which returns a list that can be handed directly to the `rix()` `git_pkgs` argument - like this: https://github.com/RichardJActon/rix/commit/68ebf8adcf8f9568942b0552f31558b1a66fcff4 because I don't really see what can be abstracted out of `renv_remote_pkg()` another function would still need the individual package info, the type, the validation of the type and the logic to change behavior based on the type and that's all that's in `renv_remote_pkg()`?

Reply:
Right, the idea is to have all this generated in one place by one function

Reply:
I've consolidated this into one function and added some error handling, and some more tests.